### PR TITLE
Rewrite Node to not use Option

### LINKED ISSUE #7: Implement Rope::insert family of methods
Depends on #6

### LINKED ISSUE #16: Do we even need to use Option in Rope?
There's been some debate between @cjm00 and I in the past over whether it was better to use `Option<Box<Node>>` for representing empty nodes, or have a `Node::Nil` variant.

Regardless of the pros & cons of those approaches, it occurs to me now that we may not actually need _either_. Empty nodes occur very rarely. Mostly, we end up with empty nodes when we've created a brand new `Rope` with no contents, or when all the text is stored in one `Leaf` node. An empty node, then, is essentially just a `Branch` node without a `String` stored in it. Why not just represent empty nodes as `Leaf("")`? 

This would remove a lot of unnecessary boilerplate for dealing with `Option`s  which are almost always `Some`. In a majority of the cases where we need to get a value out of an `Option`, we already know that it's probably non-empty, but we have to litter our code with a lot of `if let Some(...) = self.left { ... } else { unreachable!() }` and so on. 

WDYT, @cjm00?

### PR DESCRIPTION
**Pull request by hawkw**
This PR rewrites Rope to use `Leaf("")` rather than `Option` to represent empty nodes (see #16). It also fixes some additional issues in the `Rope::*_insert_*` method family, as well as bugs in iteration in the incorrect order.

Closes #16 
Closes #7

### CODE REVIEW

**Comment by hawkw** [comment]
The only test that's still broken after all these commits is [`mutable_insert_char_test_1()`](https://github.com/hawkw/an-editor/blob/5f2deef895420c827eda15f855ef523055cf3532/an-rope/src/test.rs#L133-L139) which I have still not quite determined what the issue with is...

File: an-rope/src/internals.rs [hunk 1]
```diff
@@ -45,101 +44,74 @@ fn fibonacci(n: usize) -> usize {
...
-                left: Some(box ref mut left)
-              , ref mut right, ..
-            } = *self {
-                let mut l_leaf = mem::replace(left, Node::empty());
-                l_leaf.split(index);
-                let r = right.take().unwrap_or(box Node::empty());
-                // replacing *self with a new BranchNode will update the
-                // node's weight automagically
-                BranchNode::new(box l_leaf, r)
+            let (left, left_right) = self.left.split(index);
```
Comment (deepinthebuild) [inline an-rope/src/internals.rs:76]: This naming is sorta confusing and the comment is not very helpful, obviously `foo.split()` splits foo :)

I'd prefer it if the comments in the branches explained what the intent of the operation was, i.e. "split the left node in two new nodes: `foo` which contains blah and `bar` which contains bleh. Then make a new branch containing `foo` and `bar` as children and ..."

File: an-rope/src/internals.rs [hunk 2]
```diff
@@ -197,8 +166,7 @@ impl Node {
     fn subtree_weight (&self) -> usize {
         match self { &Leaf(ref s) => s.len()
                    , &Branch(BranchNode { ref left, .. }) =>
-                        left.as_ref().map(Box::as_ref)
-                            .map_or(0, Node::subtree_weight)
+                        left.subtree_weight()
```
Comment (deepinthebuild) [inline an-rope/src/internals.rs]: Shouldn't this be `left.len()`? Otherwise this just dives down to the leftmost leaf and returns the length of that.

File: an-rope/src/internals.rs [hunk 3]
```diff
@@ -385,3 +353,20 @@ impl ops::Add for Node {
...
+        assert!( i < len
+               , "Node::index: index {} out of bounds (length {})", i, len);
+        match *self {
+            Leaf(ref vec) => { &vec[i..i+1] }
+          , Branch(BranchNode { box ref right, .. }) if len < i =>
+                &right[i - len]
+          , Branch(BranchNode { box ref left, .. }) => &left[i]
+        }
+    }
+}
```
Comment (deepinthebuild) [inline an-rope/src/internals.rs:423]: We'll need some tests for this.

File: an-rope/src/lib.rs [hunk 4]
```diff
@@ -424,25 +425,26 @@ impl Rope {
...
     /// an_rope.prepend(Rope::from("abcd"));
-    /// assert_eq!(an_rope, "abcd");
+    /// assert_eq!(&an_rope, "abcd");
     /// ```
     /// ```
     /// use an_rope::Rope;
     /// let mut an_rope = Rope::from("abcd");
     /// an_rope.prepend(Rope::from(""));
-    /// assert_eq!(an_rope, "abcd");
+    /// assert_eq!(&an_rope, "abcd");
```
Comment (deepinthebuild) [inline an-rope/src/lib.rs]: These tests should be with non-empty `Rope`s, the trivial case is a no-op.

File: an-rope/src/lib.rs [hunk 5]
```diff
@@ -502,10 +504,10 @@ impl Rope {
...
     pub fn split(&mut self, index: usize) -> (Rope, Rope) {
-        // assert!(index <= self.len());
-        // let (l, r) = self.root.split(index);
-        // (Rope { root: l }, Rope { root: r })
-        unimplemented!()
+        assert!(index <= self.len());
+        let root = mem::replace(&mut self.root, Node::empty());
+        let (l, r) = root.split(index);
+        (Rope { root: l }, Rope { root: r })
     }
```
Comment (deepinthebuild) [inline an-rope/src/lib.rs:524]: This should just take ownership of `self` instead of leaving an empty node in place. Current behavior is quite confusing.

File: an-rope/src/lib.rs [hunk 6]
```diff
@@ -456,23 +458,23 @@ impl Rope {
...
+    /// assert_eq!(&another_rope, "abcdefgh");
     /// ```
     /// ```
     /// use an_rope::Rope;
     /// let an_rope = Rope::from("");
     /// let another_rope = an_rope.with_prepend(Rope::from("abcd"));
-    /// assert_eq!(an_rope, "");
-    /// assert_eq!(another_rope, "abcd");
+    /// assert_eq!(&an_rope, "");
+    /// assert_eq!(&another_rope, "abcd");
```
Comment (deepinthebuild) [inline an-rope/src/lib.rs:482]: Let's get rid of all of these no-op doctests.

**Review by deepinthebuild — CHANGES_REQUESTED** [review]
Looks mostly good. Fix up the tree weight calculation and add some content to the current no-op doctests and we're good.

File: an-rope/src/lib.rs (hunk 4 above)
Comment (hawkw) [reply an-rope/src/lib.rs]: I put these tests there to ensure that the empty rope wasn't an edge case that caused Weird Behaviour. Considering that we also have tests for the non-trivial case, I think they should stay.

File: an-rope/src/internals.rs (hunk 1 above)
Comment (hawkw) [reply an-rope/src/internals.rs:76]: Yeah, I think what few comments there are in this function are vestiges from a previous iteration; I'll see about making this more readable.

File: an-rope/src/internals.rs (hunk 2 above)
Comment (hawkw) [reply an-rope/src/internals.rs]: Oh, god, I thought we fixed this problem like two days ago. How did that get there?

File: an-rope/src/lib.rs (hunk 4 above)
Comment (deepinthebuild) [reply an-rope/src/lib.rs]: As discussed, let's move the no-op tests to unit tests and make the doc tests illustrate non-trivial usage.

File: an-rope/src/lib.rs (hunk 5 above)
Comment (hawkw) [reply an-rope/src/lib.rs:524]: Oh, yeah, all this `mem::replace` magic is in there just so that this method can take `&mut self`, which I think is just an artefact of an older design for splitting. This should definitely just consume `self`.

File: an-rope/src/lib.rs (hunk 4 above)
Comment (hawkw) [reply an-rope/src/lib.rs]: Done:  62e107fd57186396d785807c2389750a03cd3874

File: an-rope/src/lib.rs (hunk 5 above)
Comment (hawkw) [reply an-rope/src/lib.rs:524]: Done: e22d575c7192f602e3f22706b812e7d2ea7e0b56

File: an-rope/src/internals.rs (hunk 2 above)
Comment (hawkw) [reply an-rope/src/internals.rs]: Fixed: c04e699addacdbeca636b987fc5e30a676fae6b4

File: an-rope/src/internals.rs (hunk 1 above)
Comment (hawkw) [reply an-rope/src/internals.rs:76]: Done: 36126d9ed9c91c82623113c6775fe12562ea4c3d

**Comment by hawkw** [comment]
Okay @cjm00, I think I've made all the changes you've requested?

File: an-rope/src/internals.rs [hunk 7]
```diff
@@ -55,20 +55,38 @@ impl BranchNode {
     fn split(self, index: usize) -> (Node, Node) {
         let weight = (&self).weight;
         if index < weight {
-            // split the left node
+            // if the index to split on is less than this node's weight,
+            // we split this node's left child.
```
Comment (deepinthebuild) [inline an-rope/src/internals.rs]: This is doing the thing again where the comment describes what is going on instead of what the intent is.

File: an-rope/src/internals.rs [hunk 8]
```diff
@@ -55,20 +55,38 @@ impl BranchNode {
...
+                // if the right side of the split is empty, then the right
+                // side of the returned pair is just this node's right child
                 *self.right
             } else {
+                // otherwise, the right side of the returned pair is a branch
+                // containing the right side of the split node on the left,
+                // and this node's right child on the right
                 Node::new_branch(left_right, *self.right)
             };
             (left, right)
```
Comment (deepinthebuild) [inline an-rope/src/internals.rs:89]: Why is the return type a tuple? What does this tuple mean? Could this be a BranchNode(left, right) instead?

Comment (hawkw) [reply an-rope/src/internals.rs:89]: The return type is a tuple because, ifyou look at how `insert` works, or at `Rope::split`, the two primary consumers of this method, the new Nodes returned from `split` don't end up branched together.

In `insert`, we concat the rope to insert onto the left side of the split, and then concat the right side onto the result of that concatenation. so the left and right side of the split don't actually go in a branch together. If we returned them in a `Branch`, the caller would just move the nodes back out of that branch. We'd spend extra time boxing and then unboxing the nodes. 

Also I don't think it's semantic to put the split nodes in a branch, because the whole point of `split`ing is to separate them.

Comment (hawkw) [reply an-rope/src/internals.rs:89]: Calling `split` results in an intermediate state where something_isn't_ in the tree. Returning a `Branch ` would mask that. I think the tuple expresses the intent behind the function much better.

Comment (hawkw) [inline an-rope/src/internals.rs]: Calling `split` results in an intermediate state where something_isn't_ in the tree. Returning a `Branch ` would mask that. I think the tuple expresses the intent behind the function much better.

**Comment by hawkw** [comment]
So b10dab635c86e39f08064190760a80342ecb65c7 fixes the one remaining failing test; we're all green now!

**Comment by hawkw** [comment]
@cjm00  I know you're travelling at the moment, but this PR should be merge-able, I think?

**Comment by deepinthebuild** [comment]
Yes, go for it.
